### DISCUSSION TITLE
How to override default `files` matching?

### DISCUSSION THREAD

**Discussion by beatak** [eslint/eslint#18850] [General] [answered]
With `v 9.9.1` [the doc says](https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-and-ignores)

> Specifying files and ignores
> …By default, ESLint lints files that match the patterns `**/*.js`, `**/*.cjs`, and `**/*.mjs`. Those files are always matched unless you explicitly exclude them using ignores.

I'm runinng by `npx eslint --no-config-lookup --config eslint.config.jsonly.mjs` and its content is

```
import globals from 'globals';

export default [
  {
    languageOptions: {
      globals: {
        ...globals.node,
      }
    },
    files: ['test/**/*.js', 'bin/**/*.js'],
    ignores: ['**/*.js', '**/*.cjs', '**/*.mjs', 'tmp/*', 'dist/*'],
```

but I still get an error from `dist/*.js` because they are matched by `**/*.js`.

**Comment by amareshsm** [comment]
Try adding like this:
```
export default [
          other configs ....,
	{
         other configs.....
	},
	{
		ignores: ["**/*..js"], //ignore explicitly 
	},
}
```

**Comment by beatak** [comment, marked as answer]
Thanks @amareshsm, I didn't quite understand the idea of [`global ignores`](https://eslint.org/docs/latest/use/configure/configuration-files#globally-ignoring-files-with-ignores), but with your example, I got it. Eventually i made it like:

```js
export default [
  {
    ignores: ['**/*.mjs', '**/*.cjs', 'tmp/*', 'dist/*']
  },
  mainConfig
}
```

Now I understand that this flat config (?) can express the precedent by each block, but this `global` files/ignores doesn't feels intuitive to me. Because of such a wide inclusion (`**/*.js`) as default, the `files` property only has a meaning when you need to lint something other than `.js` files. And if you do like…

```js
export default [
  {
    ignores: ['**/*.js', '**/*.mjs', '**/*.cjs']
  },
  {
    files: ['src/**/*.js']
  },
  mainConfig
}
```

I get:

```
Oops! Something went wrong! :(

ESLint: 9.9.1

You are linting ".", but all of the files matching the glob pattern "." are ignored.

If you don't want to lint these files, remove the pattern "." from the list of arguments passed to ESLint…
```

So practically, you **cannot** override `files: ["**/*.js"]` at all. Just my 2c.

**Reply by gear54rus** [reply]
> practically

I think we can become even more practical and go for patching mechanisms (like [pnpm patch](https://pnpm.io/cli/patch) or npm [patch-package](https://www.npmjs.com/package/patch-package)) to remove the following lines:

<img width="714" alt="image" src="https://github.com/user-attachments/assets/80ecffd8-6337-44b3-915b-83e835bd2dc4" />

from `node_modules/eslint/lib/config/default-config.js` which will disable this bs.

**Comment by graue** [comment]
This confused me too. The `files` config directive doesn't do what I'd expect. I thought it meant, "lint only those files," but apparently it means, "apply my config to only those files, while linting all the other files using the default config."

As a workaround, it looks like you can pass the file pattern you want linted in the command line, something like, in your case:

    npx eslint --no-config-lookup --config eslint.config.jsonly.mjs test/**/*.js bin/**/*.js

You can also put the ESLint command in your package.json under "scripts" (the `npx` is not needed in that case), so that it's runnable with something like `npm run lint`, so you don't have to type that whole command.

**Comment by JoelVenable** [comment]
I agree, the `files` and `ignores` keys do not behave how I would expect.  I'm trying to use the new flat config in a monorepo (pnpm and turborepo), and I want fine-grained control for each package (only lint files that are explicitly matched).  

So, to override the default matching I tried something like this:

```
// <root>/eslint.config.mjs
import packageAConfig from './packages/package-a/eslint.config.mjs'

export default [
  { ignores: ['**/*'] },
  ...packageAConfig
]
```

```
// <root>/packages/package-a/eslint.config.mjs

export default [
  { 
    ignores: ['!packages/package-a/src/**/*.ts'], 
    files: ['packages/package-a/src/**/*.ts'],
    rules: { ... }
  }
]
```

Neither the `!` ignores rule nor the inclusion in files resulted in any files being matched.  Until there's a fix here, I'm downgrading to version 8, since it behaves exactly the way I wanted by simply adding an eslintrc file in each package directory.

**Reply by RafaelFerreiraTVD** [reply]
Same problem...

**Comment by ferdnyc** [comment]
I found a way to make this work.

In my project, we have certain specific subdirectories that we want to have linted. All other JS files _anywhere else_ in the tree should be ignored. "Anywhere else" can't be explicitly specified, because there's no way to know what other directories may be created within a particular instance of the repo. So, exactly the opposite of the default `'**/*.js'` glob.

This config successfully reins in ESLint 9 and "un-globs" `'**/*.js'` _without_ also ignoring all of the directories explicitly specified as non-ignored. Which got it back to being able to match the files I want it to, and _only_ the files I want it to:

```js
import js from '@eslint/js';
import {defineConfig, globalIgnores} from 'eslint/config';

export default defineConfig([
    js.configs.recommended,
    globalIgnores([
        '**/*.js',
        '!src/**/*.js',
        '!tests/**/*.js',
        '!webextension/**/*.js',
    ]),
    {
        // other config objects with files: matches like,
        files: ['src/**/*.js'],
        rules: { ... },
    }.
    {
        files: ['tests/**/*.js'],
        rules: { ... },
    }
]);
```

**Comment by valuedvision** [comment]
You need to define a global ignore list as the first array element, like this:

```
export default [
  // global ignores
  {
    ignores: ["dist/**", "node_modules/**", "bin/**", "build/**"],
  },
  {
    files: ["**/*.js", "**/*.ts"],
    rules: {
      "prefer-const": "error",
    },
  },
];

```
It's also explained over here: https://www.raulmelo.me/en/blog/migration-eslint-to-flat-config

**Reply by ferdnyc** [reply]
That was only part of the question. @beatak worked out the global ignore list. (See their accepted self-answer — which also outlines the remaining problem.)

Because of how the config is processed, `files: ["**/*.js", "**/*.ts"]` only tells ESlint what files to _apply that section of the configuration_ to. It doesn't do anything to control which files are linted at all.

The list of files that are linted now includes, _by default_, `"**/*.js"`. Even if you didn't supply any configs to customize the rules, all JS files inside the directory tree where `eslint` is run will _still_ be linted using the default rules.

My example, above (globally ignoring everything _except_ certain patterns), is the only way to _disable_ the default `"**/*.js"` glob and lint **only** the patterns explicitly defined. Which isn't something many users will need, hopefully, but it's unfortunate that it's been made so difficult for those of us who do need it.
